### ISSUE TITLE
keyword search with "heterotaxy" doesn't find OMIM terms that include "heterotaxy"

### ISSUE THREAD

**Issue by LeylaR** [intermine/intermine#1464]
A keyword search with the term "heterotaxy" did not return any OMIM disease terms that contain the word "heterotaxy". This is probably because the OMIM terms contain a comma immediately after heterotaxy (such as "Heterotaxy, visceral, 4, autosomal"). Is there anything we can do to improve this type of search? Users will expect to find these terms. 

Thanks.

**Comment by julie-sullivan** [comment]
@justinccdev Do you have comments?

@LeylaR Yes, that is a problem in HumanMine too! The issue is that we specifically chose a search parser that does _not_ ignore punctuation. This is because punctuation, like dashes, e.g. ` ZDB-GENE-990415-8` or colons, e.g. `GO:0021912`, mean something specific. Does that make sense? 

There is a simple way to ignore punctuation but that would make searching for identifiers much harder.

@sergiocontrino Did thalemine do something funny to their search? Did they add wildcards around every search automatically?

**Comment by justinccdev** [comment]
Yes, currently the InterMine code that creates the search index is hard-coded to tokenize (split strings) on whitespace but keep punctuation intact.  So I would expect "Heterotaxy," to be indexed rather than "Heterotaxy".

If you do a wildcard search with "heterotaxy*" or "heterotaxy?" does the term show up?  More documentation on lucene wildcards is at [1].

As Julie says, we need to be careful with tokenizers so that we don't do things like split IDs inappropriately.  Commas might be a special case where we could remove them from the index without ID issues (since I believe no bio IDs contain commas, though there may be insane cases).  Alternatively, it may be possible to index different fields on different criteria (e.g. just do whitespace splitting on IDs but remove punctuation from description fields).  Naturally, the latter would requirement non-trivial development work, the extent of which would depend on how sophisticated one needed to be about tokenizing different biological object fields).

[1] https://lucene.apache.org/core/3_0_3/queryparsersyntax.html

**Comment by justinccdev** [comment]
Also, if we remove to ElasticSearch/Solr for search in the future (as mentioned most recently at [1]), they may have their own way of handling these issues and the InterMine code would be removed.  However, don't know the timeline for such a shift.

[1] http://gmod.827538.n3.nabble.com/Is-it-useful-to-use-RAMDirectory-for-indexing-lucene-keywords-td4056319.html

**Comment by sergiocontrino** [comment]
@julie-sullivan in thalemine, where '?' is allowed as meta char, this is appended to the query string to avoid problems with punctuation. 
'?' [=match a single char] in lucene.
we do it only if there are more terms in the search (i.e. the query string contains >1 spaces), don't remember why...

**Comment by julie-sullivan** [comment]
@danielabutano Maybe this will be addressed with the new search?

**Comment by julie-sullivan** [comment]
It was not addressed in the new search.

Possible Solutions:

1. find a tokeniser that works. [Solr docs](https://lucene.apache.org/solr/guide/7_2/tokenizers.html) | [InterMine code](https://github.com/intermine/intermine/blob/gradle-search/intermine/api/src/main/java/org/intermine/api/searchengine/solr/SolrIndexHandler.java#L267). Maybe Arunan missed something (unlikely)?
2. write our own tokeniser.
3. update OMIM. Add "HETEROTAXY" as a synonym so that it's found in the search. Won't fix other use cases. But do we have other use cases?
4. default to a "fuzzy" search. (put * around the term) let user select "exact match".
5. Make "extend Solr search" a GSoC project. (have student do #2 and #4 and other things)

**Comment by julie-sullivan** [comment]
@sergiocontrino Did you say that thalemine fixed this specific case?
